packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart

```diff
@@ -0,0 +1,231 @@
+    );
+  }
+
+  @override
+  Future<void> unsubscribe(
+    String subscriptionId,
+  ) async {
+    logger.debug('Attempting to cancel Operation $subscriptionId');
+    _send(
+      WebSocketStopMessage(id: subscriptionId),
+    );
+  }
+
+  void _send(WebSocketMessage message) {
+    try {
+      assert(sink != null, 'Sink must exist');
+      final msgJson = json.encode(message.toJson());
+      sink!.add(msgJson);
+    } on Exception catch (e) {
+      logger.warn(
+        'Error while trying to add to sync.',
+        e,
+      );
+    }
+  }
+
+  Future<void> _sendSubscriptionRegistrationMessage<T>(
+    ConnectedState state,
+    GraphQLRequest<T> request,
+  ) async {
+    final subscriptionRegistrationMessage =
+        await generateSubscriptionRegistrationMessage(
+      state.config,
+      id: request.id,
+      authRepo: state.authProviderRepo,
+      request: request,
+    );
+    _send(subscriptionRegistrationMessage);
+  }
+
+  /// Transforms an event of <String> from AppSync and converts it into a
+  ///  [WebSocketEvent] which gets consumed in [WebSocketBloc]
+  @visibleForTesting
+  Stream<WebSocketEvent> transformStream(
+    Stream<dynamic> stream,
+  ) {
+    return stream
+        .transform(const WebSocketMessageStreamTransformer())
+        .map(_onData)
+        .whereType<WebSocketEvent>()
+        .asBroadcastStream();
+  }
+
+  WebSocketEvent? _onData(
+    WebSocketMessage message,
+  ) {
+    logger.verbose(prettyPrintJson(message));
+    switch (message.messageType) {
+      case MessageType.connectionAck:
+        final messageAck = message.payload as ConnectionAckMessagePayload;
+        return ConnectionAckMessageEvent(messageAck);
+      case MessageType.connectionError:
+        return const ConnectionErrorEvent();
+      case MessageType.keepAlive:
+        return const KeepAliveEvent();
+      case MessageType.error:
+        final wsError = message.payload as WebSocketError;
+
+        // Only handle general messages, not subscription-specific ones
+        if (message.id != null) {
+          return SubscriptionErrorEvent(
+            message.id!,
+            wsError,
+          );
+        }
+
+        final exception = ApiException(
+          'Error in GraphQL subscription.',
+          underlyingException: wsError,
+        );
+        return WsErrorEvent(exception);
+      case MessageType.startAck:
+        return SubscriptionStartAckEvent(message.id!);
+      case MessageType.data:
+        final payload = message.payload as SubscriptionDataPayload;
+        return SubscriptionDataEvent(message.id!, payload);
+      case MessageType.complete:
+        return SubscriptionComplete(message.id!);
+      // We only send these messages, so we should not be receiving them.
+      case MessageType.connectionInit:
+      case MessageType.start:
+      case MessageType.stop:
+        return null;
+    }
+  }
+
+  @override
+  Future<void> close() async {
+    await sink?.close();
+    sink = null;
+  }
```

Comment:
I would make this sync to avoid any weird race condition

```suggestion
  void close() {
    sink?.close();
    sink = null;
  }
```